enhancements/windows-containers/windows-machine-config-operator-upgrades.md

```diff
@@ -0,0 +1,191 @@
+---
+
+# Windows Machine Config Operator Upgrades
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [x] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement allows a cluster administrator to upgrade a deployed Windows 
+Machine Config Operator (WMCO), and associated Windows node components, to a
+newer version.
+
+## Motivation
+
+The main motivation behind this enhancement is to give the end user
+an easy and non-disruptive way to upgrade to a newer release of the WMCO, and
+upgrade associated Windows node components with it.
+WMCO is not part of the release payload, and is managed with [OLM](https://github.com/operator-framework/operator-lifecycle-manager).
+
+### Goals
+
+As part of this enhancement we plan to do the following:
+* Enable upgrading WMCO with as little disruption as possible.
+* Enable upgrading Kubernetes components through newer versions of the WMCO,
+  including major version upgrades across OpenShift cluster upgrades, and bug
+  and CVE fixes during the current version.
+
+### Non-Goals
+
+In this enhancement we do not plan to support Windows operating system updates.
+
+## Proposal
+
+To allow for the WMCO to be upgraded in a safe and non-disruptive way, it is
+proposed that Windows VMs configured by previous versions of the WMCO are
+indirectly terminated in batches by user created MachineHealthChecks, triggered
+by WMCO stopping the associated node's kubelet after WMCO drains said node. The
+upgraded WMCO instance will then be able to configure VMs that will be created
+in place of the terminated ones.
+
+In order to allow for this to occur with minimal service disruption, it is
+recommended that users have at least two Windows nodes within the cluster.
+
+All Windows Kubernetes component updates will be tied to WMCO releases.
+
+### Justification
+
+The reason we need to cause the VMs to be fully terminated, instead of updating
+them, is that changes made to the VM done in previous versions of the WMCO may
+cause issues with the current version. If we chose to do an in place update,
+we would need to actively undo changes made to the node that may cause issues
+with the current WMCO version. The logic to keep track and undo these changes
+adds an additional layer of complexity that also adds the benefit of a quicker
+upgrade time. The risk of introducing additional issues is not worth it at this
+time, however it should be explored as the product matures.
+
+The reason we will not handle Windows operating system updates is that the
+cluster administrator provides the Windows image to create the VMs with, and
+it stays their responsiblity to provide an updated image.
+
+It is left to the user to create the Windows MachineHealthChecks, as
```

Comment:
If that is the case should WMCO be configured to send out an alert if no `MachineHealthChecks` are not present for Windows nodes?

Reply:
Good idea, added it in the risks + mitigations section